### PR TITLE
Measure cache hits and misses

### PR DESCRIPTION
**Pull request by MyDogTom**
Attach `BuildOperationListener` to the Gradle build and measure cache hits and misses. This covers both local and remote cache. 

**How it works**
`BuildCacheOperationListener` listens for two events: `ExecuteTaskBuildOperationType` (task was executed) and `BuildCacheRemoteLoadBuildOperationType` (attempt to download remote cache). Summarised information is exposed as `ExecutedTasksInfo` entity. This information is used by `TalaiotPublisherImpl` in order to attach cache usage summary to `ExecutionReport.Environment` and to every task.

Inspired by: [BuildCacheConnectionMeasurer from gradle-doctor plugin](https://github.com/runningcode/gradle-doctor/blob/321f4d09856aa07ddb2541e205b7684c467d5779/doctor-plugin/src/main/java/com/osacky/doctor/BuildCacheConnectionMeasurer.kt)

**Automated tests**
I covered major parts with unit tests. Unfortunately, I couldn't come up with a proper e2e test. I think it's very important to have such e2e because this functionality relies on Gradle internal implementation. If you can help me and create an e2e test as a separate PR, that would be nice.

### CODE REVIEW

**Comment by cdsap** [comment]
what do you need in terms of the E2E?
With testContainers we could have a lot flexibility, even we can create a Remote Cache Instance in case is needed.

**Comment by MyDogTom** [comment]
>what do you need in terms of the E2E?

I want to verify that event `BuildCacheRemoteLoadBuildOperationType` is emitted by Gradle. Because if at some point they change internals and stop emit it, then functionality won't work anymore. It would be nice to detect it.

I guess that Remote Cache Instance is needed for that.

File: talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/TaskMetrics.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package com.cdsap.talaiot.metrics
+
+import com.cdsap.talaiot.entities.CacheInfo
+import com.cdsap.talaiot.metrics.base.ExecutedTasksMetric
+
+class CacheHitMetric : ExecutedTasksMetric<BuildCacheUsageStats>(
```
Comment (cdsap) [inline talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/TaskMetrics.kt:6]: name of the  file is TaskMetrics, can we change?

File: talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/TaskMetrics.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        var remoteCacheMissCount = 0
+        info.executedTasksInfo.values.forEach {
+            when (it.localCacheInfo) {
+                is CacheInfo.CacheHit -> localCacheHitCount++
+                is CacheInfo.CacheMiss -> localCacheMissCount++
+                CacheInfo.CacheDisabled -> {
+                }
+            }
+
+            when (it.remoteCacheInfo) {
```
Comment (cdsap) [inline talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/TaskMetrics.kt:20]: Nice! I have seen cases when there is bad connectivity Gradle disabled the cache mechanism. Is covered in the when +1

File: talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/base/TaskMetric.kt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.cdsap.talaiot.metrics.base
+
+import com.cdsap.talaiot.entities.ExecutedTasksInfo
+import com.cdsap.talaiot.entities.ExecutionReport
+
+/**
+ * [Metric] that operates on [ExecutedTasksInfo]
+ */
+abstract class ExecutedTasksMetric<T>(provider: (ExecutedTasksInfo) -> T, assigner: (ExecutionReport, T) -> Unit): Metric<T, ExecutedTasksInfo>(provider, assigner)
```
Comment (cdsap) [inline talaiot/src/main/kotlin/com/cdsap/talaiot/metrics/base/TaskMetric.kt:9]: Can we use ` ExecutedTasksMetric` as File name?

**Pull request merged**